### PR TITLE
server: throw new exception when rootdisksize is required but not set

### LINKED ISSUE #7115: Unable to deploy VM using Template and custom disk offering
##### ISSUE TYPE

 * Bug Report

##### COMPONENT NAME

~~~
VM
~~~

##### CLOUDSTACK VERSION

~~~
4.17.2.0
~~~

##### CONFIGURATION

hypervisor: xcp-ng, advanced networking

##### SUMMARY

Unable to deploy VM using Template and custom disk offering, fails with error "VM Creation failed. Volume size: -1GB is out of allowed range. Max: 1024 Min:1". This issue is similar to this issue which for ISO: https://github.com/apache/cloudstack/issues/6440 but not the same.

##### STEPS TO REPRODUCE

1) Add a new Disk Offering with customized = true
`/client/api/?isMirrored=false&name=Test+Custom+Disk+Offering&displaytext=Test+Custom+Disk+Offering&storageType=shared&cacheMode=none&provisioningType=thin&customized=true&disksizestrictness=false&command=createDiskOffering&response=json
`
2) Create a new Compute Offering with diskofferingstrictness=true (please be aware that in 4.17.2.0 there is a bug and you cannot do this via UI. There is a button for this, but it is broken. This has already being fixed https://github.com/apache/cloudstack/pull/7072) You need to add the disk offering uuid in this api call!
`/client/api/?issystem=false&name=Test+Compute+Offering&displaytext=Test+Compute+Offering&customized=false&offerha=false&limitcpuuse=false&dynamicscalingenabled=false&diskofferingid=<uuid>&cpunumber=1&cpuspeed=500&memory=512&command=createServiceOffering&diskofferingstrictness=true&response=json`

3) Go to the instance wizard in the UI and choose the just created Compute offering. Do not click anything else and go straight to “Launch instance”.

##### EXPECTED RESULTS

UI needs to pass parameter rootdisksize with a default value in header of api call deployVirtualMachine even if "Override root disk size" in Template/ISO section is disabled. I do not know where do get this default value from.

##### ACTUAL RESULTS

UI is not passing parameter rootdisksize via header when using api call deployVirtualMachine and "Override root disk size" in Template/ISO section is disabled. When enabling "Override root disk size" in Template/ISO section a default value of 10 is inserted into the wizard and it works.

### PR DESCRIPTION
**Pull request by weizhouapache**
### Description

This PR fixes #7115 

Steps to reproduce the issue
- create a service offering with "Custom Disk", "Disk offering strictness" is also enabled.
- deploy a vm with the new service offering, but not "Override root disk size"

Before this PR:
image

With this PR:
image

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial

### CODE REVIEW

File: server/src/main/java/com/cloud/vm/UserVmManagerImpl.java [hunk 1]
```diff
@@ -4320,7 +4320,14 @@ public boolean checkIfDynamicScalingCanBeEnabled(VirtualMachine vm, ServiceOffer
      */
     protected long configureCustomRootDiskSize(Map<String, String> customParameters, VMTemplateVO template, HypervisorType hypervisorType, DiskOfferingVO rootDiskOffering) {
         verifyIfHypervisorSupportsRootdiskSizeOverride(hypervisorType);
-        long rootDiskSizeInBytes = verifyAndGetDiskSize(rootDiskOffering, NumbersUtil.parseLong(customParameters.get(VmDetailConstants.ROOT_DISK_SIZE), -1));
+        Long rootDiskSizeCustomParam = null;
+        if (customParameters.containsKey(VmDetailConstants.ROOT_DISK_SIZE)) {
+            rootDiskSizeCustomParam = NumbersUtil.parseLong(customParameters.get(VmDetailConstants.ROOT_DISK_SIZE), -1);
+            if (rootDiskSizeCustomParam <= 0) {
+                throw new InvalidParameterValueException("Root disk size should be a positive number.");
```
Comment (DaanHoogland) [inline server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:4327]: this is not the text I see in the description @weizhouapache . any last minute changes?

Comment (weizhouapache) [reply server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:4327]: @DaanHoogland 
this error is thrown if root disk size is passed , but invalid.

if root disk size is required but not passed, the error is thrown by code

https://github.com/apache/cloudstack/blob/3b05797344c15a9fbd6bee4ecd38281b677f24a2/server/src/main/java/com/cloud/vm/UserVmManagerImpl.java#L4310-L4316

**Comment by harikrishna-patnala** [comment]
@weizhouapache https://github.com/apache/cloudstack/issues/7115 is this not a UI bug ?

**Comment by weizhouapache** [comment]
> @weizhouapache #7115 is this not a UI bug ?

@harikrishna-patnala 
yes, we 'd better enforce user to input the rootdisksize (in the scenario) on UI
I think the new error message is more clear than before

**Comment by weizhouapache** [comment]
@blueorangutan test

**Review by DaanHoogland — APPROVED** [review]

**Review by rohityadavcloud — APPROVED** [review]
Didn't test but LGTM - shouldn't the UI do the mandatory check?

**Comment by weizhouapache** [comment]
@blueorangutan test matrix

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
@blueorangutan test matrix

**Review by harikrishna-patnala — APPROVED** [review]
code LGTM. Early check for disk size makes sense.

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
latest smoke test looks good
```
# cat /marvin//MarvinLogs/test_deploy_vm_iso_9P5EZ4/results.txt 
Test Deploy Virtual Machine from ISO ... === TestName: test_deploy_vm_from_iso | Status : SUCCESS ===
ok

----------------------------------------------------------------------
Ran 1 test in 96.624s

OK
```

**Pull request merged**